csharp-advanced-topics/HowToUseStringPool/HowtoUseStringPoolApi/StringPoolHelper.cs

```diff
@@ -0,0 +1,223 @@
+
+    public string Init()
+    {
+        var value1 = _myPool.GetOrAdd("codemaze".AsSpan());
+        var value2 = _myPool.GetOrAdd("codemaze".AsSpan());
+
+        return string.Format("Hash of value1: {0}, value2: {1}",
+             value1.GetHashCode(), value2.GetHashCode());
+    }
+
+    public string SharedInstance(ReadOnlySpan<char> message)
+    {
+        string result;
+        if (_myPool.Size <= 10)
+            result = _myPool.GetOrAdd(message);
+        else
+            result = StringPool.Shared.GetOrAdd(message);
+
+        return result;
+    }
+
+    public bool AddUser(ReadOnlySpan<char> nameSpan, ReadOnlySpan<char> emailSpan)
+    {
+        var prefixSpan = "USER_".AsSpan();
+        var prefix = _myPool.GetOrAdd(prefixSpan);
+        var name = _myPool.GetOrAdd(nameSpan);
+
+        Span<char> combined = stackalloc char[prefixSpan.Length + nameSpan.Length];
+        prefix.CopyTo(combined);
+        name.CopyTo(combined[prefix.Length..]);
+
+        var cacheKey = _myPool.GetOrAdd(combined);
+        var cacheValue = _myPool.GetOrAdd(emailSpan);
+        _cache[cacheKey] = cacheValue;
+
+        return true;
+    }
+
+    public string GetUser(ReadOnlySpan<char> nameSpan)
+    {
+        var prefixSpan = "USER_".AsSpan();
+        var prefix = _myPool.GetOrAdd(prefixSpan);
+        var name = _myPool.GetOrAdd(nameSpan);
+
+        Span<char> combined = stackalloc char[prefixSpan.Length + nameSpan.Length];
+        prefix.CopyTo(combined);
+        name.CopyTo(combined[prefix.Length..]);
+
+        var cacheKey = _myPool.GetOrAdd(combined);
+        _cache.TryGetValue(cacheKey, out string value);
+
+        return value;
+    }
+
+    public void LogError(ReadOnlySpan<char> detailSpan)
+    {
+        var prefixSpan = "[ERROR]:".AsSpan();
+        var prefix = _myPool.GetOrAdd(prefixSpan);
+        var detail = _myPool.GetOrAdd(detailSpan);
+
+        Span<char> combined = stackalloc char[prefixSpan.Length + detailSpan.Length];
+        prefix.CopyTo(combined);
+        detail.CopyTo(combined[prefix.Length..]);
+
+        var logMessage = _myPool.GetOrAdd(combined);
+        _logger.Add(logMessage);
+    }
+
+    public string GetHostName(string url)
+    {
+        int offset = url.AsSpan().IndexOf([':', '/', '/']);
+        int start = offset == -1 ? 0 : offset + 3;
+        int end = url.AsSpan(start).IndexOf('/');
+
+        ReadOnlySpan<char> hostName = url.AsSpan(start, end);
+        var result = _myPool.GetOrAdd(hostName);
+
+        return result;
+    }
+
+    public string GetHeaderValue(HttpRequest request, ReadOnlySpan<char> key)
+    {
+        string keyValue = _myPool.GetOrAdd(key);
+
+        if (request.Headers.ContainsKey(keyValue))
+            return _myPool.GetOrAdd(request.Headers[keyValue]);
+
+        return string.Empty;
+    }
+
+    public bool CheckHeader(HttpRequest request)
+    {
+        string authorization = GetHeaderValue(request, "Authorization".AsSpan());
```

Comment:
string are implicitly convertable to ReadOnlySpan<char> so no need for the "AsSpan" before passing a string to a method that takes a ReadOnlySpan<char>

Note: I just chose this one, there are several in the PR, so please go through it and address them.

Reply:
Corrected